src/mesh_tools.rs

```diff
@@ -5,8 +5,161 @@ use nalgebra::geometry::Point3;
+        let mut current_test_edge: OrientedEdge =
+            original_triangle_faces[cast_usize(current_face_index)].to_oriented_edges()[0];
+
+        // Put the first edge-face couple to the stack
+        edge_face_stack.push((current_test_edge, current_face_index));
+
+        // Check and revert (if needed) faces in the stack as long as there is any
+        while !edge_face_stack.is_empty() {
+            // Get ready for the next iteration.
+            let next_edge_face = edge_face_stack.pop().expect("Popping from an empty vector");
+            current_test_edge = next_edge_face.0;
+            current_face_index = next_edge_face.1;
+
+            let current_face = original_triangle_faces[cast_usize(current_face_index)];
+            // Use if contains reverted edge, if not, revert and use. It should be
+            // safe to presume the face contains the edge one way or another because
+            // it's given by the topology generator.
+            let proper_current_face = if current_face.contains_oriented_edge(current_test_edge) {
+                current_face
+            } else {
+                current_face.to_reverted()
+            };
+            synchronized_triangle_faces.push(proper_current_face);
+
+            // Find the indices of edges of the current face in the list of edges,
+            // from which the topology was created
+            let face_unoriented_edges = proper_current_face.to_unoriented_edges();
+            let face_unoriented_edge_indices =
+                face_unoriented_edges.iter().map(|unoriented_edge| {
+                    unoriented_edge_index_map
+                        .get(unoriented_edge)
+                        .expect("The current edge is not found in the edge collection")
+                });
+
+            // Convert into oriented edges so that the neighboring faces can check
+            // for correct winding
+            let face_oriented_edges: Vec<_> =
+                face_unoriented_edges.iter().map(|u_e| u_e.0).collect();
+            // For each face edge index
+            for (i, face_unoriented_edge_index) in face_unoriented_edge_indices.enumerate() {
+                // get the actual oriented edge
+                let face_oriented_edge = face_oriented_edges[i];
+                // and try to find it in the edge-to-face topology.
+                if let Some(edge_in_faces) = edge_to_face_topology.get(&face_unoriented_edge_index)
+                {
+                    // If it exists, iterate the faces containing the edge
+                    for face_index in edge_in_faces {
+                        // and if it was not already added to the stack or even checked
+                        if !face_processed_pattern[cast_usize(*face_index)] {
+                            // add it to the stack with the expected edge orientation
+                            edge_face_stack.push((face_oriented_edge.to_reverted(), *face_index));
```

Comment:
It might be the lateness of the hour, but why do we revert the edge here?

Reply:
because we are expected to find the reverted edge in the neighboring face. i'm reverting the face already here rather than when checking the neighbor to keep related data in the newly created tuple: it's the edge to be found and the face where it's supposed to be found